File: SiliconLabsApp/ViewControllers/BluetoothBrowser/Utils/SILBluetoothBrowserExpandableViewManager.m
//
//  SILBluetoothBrowserExpandableViewManager.m
//  BlueGecko
//
//  Created by Kamil Czajka on 26/03/2020.
//  Copyright © 2020 SiliconLabs. All rights reserved.
//

#import <Foundation/Foundation.h>
#import <UIKit/UIKit.h>
#import "SILBluetoothBrowserExpandableViewManager.h"
#import "UIImage+SILImages.h"
#import "SILStoryboard+Constants.h"
#import "SILBrowserLogViewController.h"
#import "SILBrowserConnectionsViewController.h"
#import "SILBrowserFilterViewController.h"
#import "SILBluetoothBrowser+Constants.h"

@interface SILBluetoothBrowserExpandableViewManager ()

@property (strong, nonatomic) UIButton* logButton;
@property (strong, nonatomic) UIButton* connectionsButton;
@property (strong, nonatomic) UIButton* filterButton;
@property (strong, nonatomic) UIImageView* activeFilterImageView;
@property (strong, nonatomic) NSLayoutConstraint* expandableControllerHeight;
@property (strong, nonatomic) UIView* expandableControllerView;
@property (strong, nonatomic) UIViewController* expandingViewController;
@property (strong, nonatomic) UIVisualEffectView* effectView;
@property (strong, nonatomic) UIView* presentationView;
@property (strong, nonatomic) UIView* discoveredDevicesView;
@property (strong, nonatomic) UIViewController* browserViewController;
@property (nonatomic, assign) CGFloat cornerRadius;

@end

@implementation SILBluetoothBrowserExpandableViewManager

- (instancetype)initWithOwnerViewController:(UIViewController*)viewController {
    self = [super init];
    if (self) {
        _browserViewController = viewController;
        _cornerRadius = 0;
    }
    return self;
}

- (void)setupButtonsTabBarWithLog:(UIButton*)logButton connections:(UIButton*)connectionsButton {
    self.logButton = logButton;
    self.connectionsButton = connectionsButton;
    [self setupLogButton];
    [self setupConnectionButton];
}

- (void)setupButtonsTabBarWithLog:(UIButton*)logButton connections:(UIButton*)connectionsButton filter:(UIButton*)filterButton andActiveFilterImage:(UIImageView*)activeImageView {
    self.logButton = logButton;
    self.connectionsButton = connectionsButton;
    self.filterButton = filterButton;
    self.activeFilterImageView = activeImageView;
    [self setupLogButton];
    [self setupConnectionButton];
    [self setupFilterButton];
}

- (void)setReferenceForPresentationView:(UIView*)presentationView andDiscoveredDevicesView:(UIView*)discoveredDevicesView {
    self.presentationView = presentationView;
    self.discoveredDevicesView = discoveredDevicesView;
}

- (void)setReferenceForExpandableControllerView:(UIView*)expandableControllerView andExpandableControllerHeight:(NSLayoutConstraint*)expandableControllerHeight {
    self.expandableControllerView = expandableControllerView;
    self.expandableControllerHeight = expandableControllerHeight;
}

- (void)setValueForCornerRadius:(CGFloat)cornerRadius {
    self.cornerRadius = cornerRadius;
}

- (SILBrowserLogViewController*)logButtonWasTappedAction {
    SILBrowserLogViewController* logVC;
    
    if (self.logButton.isSelected == NO) {
        BOOL anyButtonSelected = [self isAnyButtonSelected];
        [self prepareSceneDependOnButtonSelection:anyButtonSelected];
        
        UIStoryboard* storyboard = [UIStoryboard storyboardWithName:SILAppBluetoothBrowserHome bundle:nil];
        logVC = [storyboard instantiateViewControllerWithIdentifier:SILSceneLog];
    
        [self insertIntoContainerExpandableController:logVC];
        [self animateExpandableViewControllerIfNeeded:anyButtonSelected];
        self.expandingViewController = logVC;
        
        [self.logButton setSelected:YES];
    } else {
        [self prepareSceneForRemoveExpandingController];
    }
    
    return logVC;
}

- (SILBrowserConnectionsViewController*)connectionsButtonWasTappedAction {
    SILBrowserConnectionsViewController* connectionVC;
    
    if (self.connectionsButton.isSelected == NO) {
        BOOL anyButtonSelected = [self isAnyButtonSelected];
        [self prepareSceneDependOnButtonSelection:anyButtonSelected];
    
        UIStoryboard* storyboard = [UIStoryboard storyboardWithName:SILAppBluetoothBrowserHome bundle:nil];
        connectionVC = [storyboard instantiateViewControllerWithIdentifier:SILSceneConnections];
    
        [self insertIntoContainerExpandableController:connectionVC];
        [self animateExpandableViewControllerIfNeeded:anyButtonSelected];
        self.expandingViewController = connectionVC;
        
        [self.connectionsButton setSelected:YES];
    } else {
        [self prepareSceneForRemoveExpandingController];
    }
    
    return connectionVC;
}

- (SILBrowserFilterViewController*)filterButtonWasTappedAction {
    SILBrowserFilterViewController* filterVC;
    
    if (self.filterButton.isSelected == NO) {
        BOOL anyButtonSelected = [self isAnyButtonSelected];
        [self prepareSceneDependOnButtonSelection:anyButtonSelected];
        
        UIStoryboard* storyboard = [UIStoryboard storyboardWithName:SILAppBluetoothBrowserHome bundle:nil];
        filterVC = [storyboard instantiateViewControllerWithIdentifier:SILSceneFilter];
    
        [self insertIntoContainerExpandableController:filterVC];
        [self animateExpandableViewControllerIfNeeded:anyButtonSelected];
        self.expandingViewController = filterVC;
        
        [self.filterButton setSelected:YES];
    } else {
        [self prepareSceneForRemoveExpandingController];
    }
    
    return filterVC;
}

- (void)removeExpandingControllerIfNeeded {
    [self prepareSceneForRemoveExpandingController];
}

- (void)updateConnectionsButtonTitle:(NSUInteger)connections {
    NSString* const AppendingConnections = @" Connections";
    NSString* connectionsText = [NSString stringWithFormat:@"%lu%@", (unsigned long)connections, AppendingConnections];
    [self.connectionsButton setTitle:connectionsText forState:UIControlStateNormal];
    [self.connectionsButton setTitle:connectionsText forState:UIControlStateSelected];
}

#pragma mark - Private Functions

- (void)setupLogButton {
    UIEdgeInsets const ImageInsetsForLogButton = {0, 16, 0 ,8};
    UIEdgeInsets const TitleEdgeInsetsForLogButton = {0, 20., 0, 0};
    
    [self.logButton setTintColor:[UIColor clearColor]];
    [self.logButton setImage:[[UIImage imageNamed:SILImageLogOff] imageWithRenderingMode:UIImageRenderingModeAlwaysOriginal] forState: UIControlStateNormal];
    [self.logButton setImage:[[UIImage imageNamed:SILImageLogOn] imageWithRenderingMode:UIImageRenderingModeAlwaysOriginal] forState:UIControlStateSelected];
    self.logButton.imageView.contentMode = UIViewContentModeScaleAspectFit;
    self.logButton.imageEdgeInsets = ImageInsetsForLogButton;
    [self.logButton setTitleEdgeInsets:TitleEdgeInsetsForLogButton];
    self.logButton.titleLabel.adjustsFontSizeToFitWidth = YES;
    self.logButton.titleLabel.font = [UIFont robotoMediumWithSize:[UIFont getMiddleFontSize]];
    [self.logButton setTitleColor:[UIColor sil_primaryTextColor] forState:UIControlStateNormal];
    [self.logButton setTitleColor:[UIColor sil_regularBlueColor] forState:UIControlStateSelected];
}

- (void)setupConnectionButton {
    UIEdgeInsets const ImageInsetsForConnectionsButton = {0, 8, 0 ,8};
    UIEdgeInsets const TitleEdgeInsetsForConnectionsButton = {0, 8, 0, 0};
    
    [self.connectionsButton setTintColor:[UIColor clearColor]];
    [self.connectionsButton setImage:[[UIImage imageNamed:SILImageConnectOff] imageWithRenderingMode:UIImageRenderingModeAlwaysOriginal] forState:UIControlStateNormal];
    [self.connectionsButton setImage:[[UIImage imageNamed:SILImageConnectOn] imageWithRenderingMode:UIImageRenderingModeAlwaysOriginal] forState:UIControlStateSelected];
    self.connectionsButton.imageView.contentMode = UIViewContentModeScaleAspectFit;
    self.connectionsButton.imageEdgeInsets = ImageInsetsForConnectionsButton;
    [self.connectionsButton setTitleEdgeInsets:TitleEdgeInsetsForConnectionsButton];
    self.connectionsButton.titleLabel.adjustsFontSizeToFitWidth = YES;
    self.connectionsButton.titleLabel.font = [UIFont robotoMediumWithSize:[UIFont getMiddleFontSize]];
    [self.connectionsButton setTitleColor:[UIColor sil_primaryTextColor] forState:UIControlStateNormal];
    [self.connectionsButton setTitleColor:[UIColor sil_regularBlueColor] forState:UIControlStateSelected];
}

- (void)setupFilterButton {
    UIEdgeInsets const ImageInsetsForFilterButton = {0, 8, 0 ,12};
    UIEdgeInsets const TitleEdgeInsetsForFilterButton = {0, 8, 0, 8};
    
    [self.filterButton setTintColor:[UIColor clearColor]];
    [self.filterButton setImage:[[UIImage imageNamed:SILImageSearchOff] imageWithRenderingMode:UIImageRenderingModeAlwaysOriginal] forState:UIControlStateNormal];
    [self.filterButton setImage:[[UIImage imageNamed:SILImageSearchOn] imageWithRenderingMode:UIImageRenderingModeAlwaysOriginal] forState:UIControlStateSelected];
    self.filterButton.imageView.contentMode = UIViewContentModeScaleAspectFit;
    self.filterButton.imageEdgeInsets = ImageInsetsForFilterButton;
    [self.filterButton setTitleEdgeInsets:TitleEdgeInsetsForFilterButton];
    self.filterButton.titleLabel.adjustsFontSizeToFitWidth = YES;
    self.filterButton.titleLabel.font = [UIFont robotoMediumWithSize:[UIFont getMiddleFontSize]];
    [self.filterButton setTitleColor:[UIColor sil_primaryTextColor] forState:UIControlStateNormal];
    [self.filterButton setTitleColor:[UIColor sil_regularBlueColor] forState:UIControlStateSelected];
    [self.activeFilterImageView setHidden:YES];
}

- (BOOL)isAnyButtonSelected {
    return ! (self.logButton.isSelected == NO && self.connectionsButton.isSelected == NO && self.filterButton.isSelected == NO);
}

- (void)deselectAllButtons {
    [self.logButton setSelected:NO];
    [self.connectionsButton setSelected:NO];
    [self.filterButton setSelected:NO];
}

- (void)prepareSceneDependOnButtonSelection:(BOOL)anyButtonSelected {
    if (anyButtonSelected) {
        [self prepareSceneForChangeExpandableView];
    } else {
        [self prepareSceneForExpandableView];
    }
    
    [self customizeExpandableViewAppearance];
}

- (void)prepareSceneForChangeExpandableView {
    [self deselectAllButtons];
    [self removeExpandableViewController];
}

- (void)prepareSceneForExpandableView {
    if (self.expandingViewController != nil) {
        [self prepareSceneForRemoveExpandingController];
    }
    
    [self attachBlurEffectView];
}

- (void)customizeExpandableViewAppearance {
    self.cornerRadius = 20.0;
    self.expandableControllerHeight.constant = self.presentationView.frame.size.height * 0.9;
}

- (void)customizeSceneWithoutExpandableViewContoller {
    self.cornerRadius = 0.0;
    self.expandableControllerHeight.constant = CollapsedViewHeight;
}

- (void)removeExpandableViewController {
    [self.browserViewController willMoveToParentViewController:nil];
    [self.expandingViewController.view removeFromSuperview];
    [self.expandingViewController removeFromParentViewController];
    self.expandingViewController = nil;
}

- (void)insertIntoContainerExpandableController:(UIViewController*)viewController {
    [self.browserViewController addChildViewController:viewController];
    [self.expandableControllerView addSubview:viewController.view];
    viewController.view.frame = self.expandableControllerView.frame;
    viewController.view.autoresizingMask = UIViewAutoresizingFlexibleHeight | UIViewAutoresizingFlexibleWidth;
    [viewController didMoveToParentViewController:self.browserViewController];
    
    [self.browserViewController.view setNeedsUpdateConstraints];
}

- (void)animateExpandableViewControllerIfNeeded:(BOOL)anyButtonSelected {
    if (!anyButtonSelected) {
        [self animateExpandableViewController];
    }
}

- (void)animateExpandableViewController {
    [UIView animateWithDuration:AnimationExpandableControllerTime delay:AnimationExpandableControllerDelay options:UIViewAnimationOptionCurveLinear | UIViewAnimationOptionTransitionCurlDown animations:^{
        [self.browserViewController.view layoutIfNeeded];
    } completion:nil];
}

- (void)prepareSceneForRemoveExpandingController {
    [self deselectAllButtons];
    [self removeExpandableViewController];
    [self customizeSceneWithoutExpandableViewContoller];
    [self animateExpandableViewController];
    [self removeBlurEffectView];
}

- (void)attachBlurEffectView {
    UIBlurEffect* blurEffect = [UIBlurEffect effectWithStyle:UIBlurEffectStyleDark];
    self.effectView = [[UIVisualEffectView alloc] initWithEffect:blurEffect];
    self.effectView.frame = self.presentationView.frame;
    [self.discoveredDevicesView addSubview:self.effectView];
}

- (void)removeBlurEffectView {
    [self.effectView removeFromSuperview];
}

- (void)setCornerRadius:(CGFloat)cornerRadius {
    if (_cornerRadius != cornerRadius) {
        _cornerRadius = cornerRadius;
        [self adjustView];
    }
}

- (void)adjustView {
    if (self.expandableControllerView != nil) {
        self.expandableControllerView.layer.cornerRadius = self.cornerRadius;
        self.expandableControllerView.layer.maskedCorners = kCALayerMaxXMaxYCorner | kCALayerMinXMaxYCorner;
        self.expandableControllerView.clipsToBounds = YES;
    }
}

@end

File: SiliconLabsApp/ViewControllers/BluetoothBrowser/Utils/SILBluetoothBrowserExpandableViewManager.swift
//
//  SILBluetoothBrowserExpandableViewManager.swift
//  BlueGecko
//
//  Created by Grzegorz Janosz on 18/11/2020.
//  Copyright © 2020 SiliconLabs. All rights reserved.
//

@objc
@objcMembers
class SILBluetoothBrowserExpandableViewManager: NSObject {
    
    private var logButton: UIButton?
    private var connectionsButton: UIButton?
    private var filterButton: UIButton?
    private var sortButton: UIButton?
    private var expandableControllerHeight: NSLayoutConstraint?
    private var expandableControllerView: UIView?
    private var expandingViewController: UIViewController?
    private var filterBarHeight: NSLayoutConstraint?
    private var filterBarViewController: SILFilterBarViewController?
    
    private var effectView = UIVisualEffectView(effect: UIBlurEffect(style: .dark))
    
    private var presentationView: UIView?
    private var discoveredDevicesView: UIView?
    private var browserViewController: UIViewController?
    private var cornerRadius: CGFloat = 0 {
        didSet {
            if  cornerRadius != oldValue {
                adjustView()
            }
        }
    }
    
    private var usedButtons: [UIButton] = []
    
    var filterIsSelected = false
    
    init(withOwnerViewController viewController: UIViewController) {
        super.init()
        self.browserViewController = viewController
    }
    
    // MARK: Setup
    
    func setReferenceFor(presentationView: UIView, andDiscoveredDevicesView discoveredDevicesView: UIView) {
        self.presentationView = presentationView
        self.discoveredDevicesView = discoveredDevicesView
    }
    
    
    func setReferenceForExpandableControllerView(_ expandableControllerView: UIView, andExpandableControllerHeight expandableControllerHeight: NSLayoutConstraint) {
        self.expandableControllerView = expandableControllerView
        self.expandableControllerHeight = expandableControllerHeight
    }
    
    func setValueFor(cornerRadius: CGFloat) {
        self.cornerRadius = cornerRadius
    }
    
    func setupButtonsTabBar(log logButton: UIButton, connections connectionsButton: UIButton) {
        self.logButton = logButton
        self.connectionsButton = connectionsButton
        [logButton, connectionsButton].forEach {
            usedButtons.append($0)
        }
        setupConnectionButton()
    }
    
    func setupButtonsTabBar(log logButton: UIButton, connections connectionsButton: UIButton, filter filterButton: UIButton, andFilterIsActive isActive: Bool, andSortButton sortButton: UIButton) {
        self.logButton = logButton
        self.connectionsButton = connectionsButton
        self.filterButton = filterButton
        self.sortButton = sortButton
        [logButton, connectionsButton, filterButton, sortButton].forEach {
            usedButtons.append($0)
        }
        setupConnectionButton()
        setupFilterButtonWhereIsFilterActive(isActive)
        setupSortButton()
    }

    private func setupConnectionButton() {
        connectionsButton?.setImage(UIImage(named: SILImageConnectOff)!.withRenderingMode(.alwaysOriginal), for: .normal)
        connectionsButton?.setImage(UIImage(named: SILImageConnectOn)!.withRenderingMode(.alwaysOriginal), for: .selected)
    }
    
    private func setupFilterButtonWhereIsFilterActive(_ isFilterActive: Bool) {
        updateFilterIsActiveFilter(isFilterActive)
    }
    
    private func setupSortButton() {
        sortButton?.setImage(UIImage(named: SILImageSortOff)!.withRenderingMode(.alwaysOriginal), for: .normal)
        sortButton?.setImage(UIImage(named: SILImageSortOn)!.withRenderingMode(.alwaysOriginal), for: .selected)
    }
    
    private func setupButton(_ button: UIButton) {
        let titleEdgeInsetsForButton = UIEdgeInsets(top: 0, left: 7, bottom: 0, right: 0)
        button.tintColor = .clear
        button.imageView?.contentMode = .scaleAspectFit
        button.titleEdgeInsets = titleEdgeInsetsForButton
        button.titleLabel?.font = UIFont.robotoMedium(size: UIFont.getMiddleFontSize())
        button.setTitleColor(UIColor.sil_primaryText(), for: .normal)
        button.setTitleColor(UIColor.sil_regularBlue(), for: .selected)
    }
    
    func setupFilterBar(filterBarHeight: NSLayoutConstraint?, filterBarViewController: SILFilterBarViewController?) {
        self.filterBarHeight = filterBarHeight
        self.filterBarViewController = filterBarViewController
    }
    
    func updateFilterIsActiveFilter(_ isActiveFilter: Bool) {
        if isActiveFilter {
            setupImagesForActiveFilter()
        } else {
            setupImagesForInactiveFilter()
        }
    }
    
    func updateFilterBar(with filterViewModel: SILBrowserFilterViewModel?) {
        if let filterViewModel = filterViewModel {
            self.filterBarViewController?.updateCurrentFilter(filter: filterViewModel)
            self.filterBarHeight?.isActive = !filterViewModel.isFilterActive()
        } else {
           hideFilterBar()
        }
    }
    
    private func hideFilterBar() {
        self.filterBarHeight?.isActive = true
    }
    
    private func setupImagesForInactiveFilter() {
        filterButton?.setImage(UIImage(named: SILImageFilterOff)!.withRenderingMode(.alwaysOriginal), for: .normal)
        filterButton?.setImage(UIImage(named: SILImageFilterOffSelected)!.withRenderingMode(.alwaysOriginal), for: .selected)
    }

    private func setupImagesForActiveFilter() {
        filterButton?.setImage(UIImage(named: SILImageFilterOn)!.withRenderingMode(.alwaysOriginal), for: .normal)
        filterButton?.setImage(UIImage(named: SILImageFilterOnSelected)!.withRenderingMode(.alwaysOriginal), for: .selected)
    }

    func changeImagesOfSortButtonFor(option: SILSortOption) {
        switch option {
        case .none:
            sortButton?.setImage(UIImage(named: SILImageSortOff)!.withRenderingMode(.alwaysOriginal), for: .normal)
            sortButton?.setImage(UIImage(named: SILImageSortOn)!.withRenderingMode(.alwaysOriginal), for: .selected)
        case .ascendingRSSI:
            sortButton?.setImage(UIImage(named: SILImageSortAscendingOff)!.withRenderingMode(.alwaysOriginal), for: .normal)
            sortButton?.setImage(UIImage(named: SILImageSortAscendingOn)!.withRenderingMode(.alwaysOriginal), for: .selected)
        case .descendingRSSI:
            sortButton?.setImage(UIImage(named: SILImageSortDescendingOff)!.withRenderingMode(.alwaysOriginal), for: .normal)
            sortButton?.setImage(UIImage(named: SILImageSortDescendingOn)!.withRenderingMode(.alwaysOriginal), for: .selected)
        case .AZ:
            sortButton?.setImage(UIImage(named: SILImageSortAZOff)!.withRenderingMode(.alwaysOriginal), for: .normal)
            sortButton?.setImage(UIImage(named: SILImageSortAZOn)!.withRenderingMode(.alwaysOriginal), for: .selected)
        case .ZA:
            sortButton?.setImage(UIImage(named: SILImageSortZAOff)!.withRenderingMode(.alwaysOriginal), for: .normal)
            sortButton?.setImage(UIImage(named: SILImageSortZAOn)!.withRenderingMode(.alwaysOriginal), for: .selected)
        }
    }
    
    // MARK: TappedActions
    
    func logButtonWasTappedAction() -> SILBrowserLogViewController? {
        var logVC: SILBrowserLogViewController? = nil
        
        if !logButton!.isSelected {
            let storyboard = UIStoryboard(name: SILAppBluetoothBrowserHome, bundle: nil)
            logVC = storyboard.instantiateViewController(withIdentifier: SILSceneLog) as? SILBrowserLogViewController
            handleButtonSelect(self.logButton!, andViewController: logVC!)
        } else {
            prepareSceneForRemoveExpandingController()
        }
        
        return logVC
    }
    
    func connectionsButtonWasTappedAction() -> SILBrowserConnectionsViewController? {
        var connectionVC: SILBrowserConnectionsViewController? = nil
        
        if !connectionsButton!.isSelected {
            let storyboard = UIStoryboard(name: SILAppBluetoothBrowserHome, bundle: nil)
            connectionVC = storyboard.instantiateViewController(withIdentifier: SILSceneConnections) as? SILBrowserConnectionsViewController
            handleButtonSelect(self.connectionsButton!, andViewController: connectionVC!)
        } else {
            prepareSceneForRemoveExpandingController()
        }
        
        return connectionVC
    }
    
    func filterButtonWasTappedAction() -> SILBrowserFilterViewController? {
        var filterVC: SILBrowserFilterViewController? = nil
        
        if filterIsSelected {
            let storyboard = UIStoryboard(name: SILAppBluetoothBrowserHome, bundle: nil)
            filterVC = storyboard.instantiateViewController(withIdentifier: SILSceneFilter) as? SILBrowserFilterViewController
            handleButtonSelect(UIButton(), andViewController: filterVC!)
        } else {
            prepareSceneForRemoveExpandingController()
        }

        return filterVC
    }
    
    private func handleButtonSelect(_ button: UIButton, andViewController vc: UIViewController, wasSortButtonSelected: Bool = false) {
        let wasAnyButtonSelected = isAnyButtonSelected()
        let wasSortButtonSelected = sortButton?.isSelected ?? false
        let willBeSortButtonSelected = button.isEqual(self.sortButton)
        deselectAllButtons()
        button.isSelected = true
        prepareSceneDependOnButtonSelection(wasAnyButtonSelected)

        insertIntoContainerExpandableController(vc)
        let isNeededAnimate = wasSortButtonSelected || !wasAnyButtonSelected || willBeSortButtonSelected
        animateExpandableViewController(isNeeded: isNeededAnimate)
        self.expandingViewController = vc
    }
    
    // MARK: Public appearance methods
    
    func updateConnectionsButtonTitle(_ connections: UInt) {
        let connectionText = "\(connections) Connections"
        connectionsButton?.setTitle(connectionText, for: .normal)
        connectionsButton?.setTitle(connectionText, for: .selected)
    }
    
    func removeExpandingControllerIfNeeded() {
        prepareSceneForRemoveExpandingController()
    }
    
    // MARK: Private appearance methods
    
    private func isAnyButtonSelected() -> Bool {
        return usedButtons.reduce(false) { $0 || $1.isSelected }
    }
    
    private func deselectAllButtons() {
        usedButtons.forEach{ $0.isSelected = false }
    }
    
    private func prepareSceneDependOnButtonSelection(_ wasAnyButtonSelected: Bool) {
        if wasAnyButtonSelected {
            prepareSceneForChangeExpandableView()
        } else {
            prepareSceneForExpandableView()
        }
        customizeExpandableViewAppearance()
    }
    
    private func insertIntoContainerExpandableController(_ viewController: UIViewController) {
        browserViewController!.addChild(viewController)
        expandableControllerView!.addSubview(viewController.view)
        viewController.view.frame = self.expandableControllerView!.frame
        viewController.view.autoresizingMask = [.flexibleHeight, .flexibleWidth]
        viewController.didMove(toParent: self.browserViewController!)
        browserViewController!.view.setNeedsUpdateConstraints()
    }
    
    private func prepareSceneForChangeExpandableView() {
        removeExpandableViewController()
    }
    
    private func prepareSceneForExpandableView() {
        if let _ = self.expandingViewController {
            prepareSceneForRemoveExpandingController()
        }
        hideFilterBar()
        attachBlurEffectView()
    }
    
    private func customizeExpandableViewAppearance() {
        self.cornerRadius = 20.0
        if sortButton?.isSelected ?? false {
            let sortVM = SILSortViewModel._sharedInstance
            expandableControllerHeight!.constant = sortVM.getViewControllerHeight()
            return
        }
        self.expandableControllerHeight!.constant = self.presentationView!.frame.size.height * 0.9
    }
    
    private func customizeSceneWithoutExpandableViewContoller() {
        self.cornerRadius = 0.0
        self.expandableControllerHeight?.constant = CollapsedViewHeight
    }
    
    private func removeExpandableViewController() {
        browserViewController?.willMove(toParent: nil)
        expandingViewController?.view.removeFromSuperview()
        expandingViewController?.removeFromParent()
        expandingViewController = nil
    }
    
    private func prepareSceneForRemoveExpandingController() {
        deselectAllButtons()
        removeExpandableViewController()
        customizeSceneWithoutExpandableViewContoller()
        animateExpandableViewController()
        removeBlurEffectView()
        restoreFilterBar()
        self.filterIsSelected = false
    }
    
    private func restoreFilterBar() {
        if let filterBarViewController = self.filterBarViewController,
           !filterBarViewController.isEmpty() {
            filterBarViewController.restore()
            self.filterBarHeight?.isActive = false
        }
    }
        
    private func attachBlurEffectView() {
        self.effectView.frame = self.presentationView!.frame
        self.discoveredDevicesView?.addSubview(self.effectView)
    }

    private func removeBlurEffectView() {
        self.effectView.removeFromSuperview()
    }

    private func adjustView() {
        if let expandableView = self.expandableControllerView {
            expandableView.layer.cornerRadius = self.cornerRadius
            expandableView.layer.maskedCorners = [.layerMinXMaxYCorner, .layerMaxXMaxYCorner]
            expandableView.clipsToBounds = true
        }
    }
    
    private func animateExpandableViewController(isNeeded: Bool = true) {
        if isNeeded {
            UIView.animate(withDuration: TimeInterval(AnimationExpandableControllerTime), delay: TimeInterval(AnimationExpandableControllerDelay), options: [.curveLinear], animations: { self.browserViewController?.view.layoutIfNeeded()
            }, completion: nil)
        }
    }
}
